improver_tests/calibration/dataframe_utilities/test_dataframe_utilities.py

```diff
@@ -638,6 +638,74 @@ def test_truth_missing_compulsory_columns(self):
                 self.training_length,
             )
 
+    def test_duplicate_cycle_forecasts(self):
+        """Test that a forecast cube is still produced if a duplicated
+        cycle of forecasts is provided."""
+        forecast_df_with_duplicates = self.forecast_df.append(
+            self.forecast_df.iloc[:9], ignore_index=True
+        )
+        result = forecast_and_truth_dataframes_to_cubes(
+            forecast_df_with_duplicates,
+            self.truth_subset_df,
+            self.cycletime,
+            self.forecast_period,
+            self.training_length,
+        )
+        self.assertEqual(len(result), 2)
+        self.assertCubeEqual(result[0], self.expected_period_forecast)
+        self.assertCubeEqual(result[1], self.expected_period_truth)
+
+    def test_duplicate_cycle_truths(self):
+        """Test that a truth cube is still produced if duplicate
+        truths for a given validity time are provided."""
+        truth_df_with_duplicates = self.truth_subset_df.append(
+            self.truth_subset_df.iloc[:3], ignore_index=True
+        )
+        result = forecast_and_truth_dataframes_to_cubes(
+            self.forecast_df,
+            truth_df_with_duplicates,
+            self.cycletime,
+            self.forecast_period,
+            self.training_length,
+        )
+        self.assertEqual(len(result), 2)
+        self.assertCubeEqual(result[0], self.expected_period_forecast)
+        self.assertCubeEqual(result[1], self.expected_period_truth)
+
+    def test_duplicate_row_forecasts(self):
+        """Test that a forecast cube is still produced if a single
+        duplicated forecast is provided."""
+        forecast_df_with_duplicates = self.forecast_df.append(
+            self.forecast_df.iloc[0], ignore_index=True
+        )
+        result = forecast_and_truth_dataframes_to_cubes(
+            forecast_df_with_duplicates,
+            self.truth_subset_df,
+            self.cycletime,
+            self.forecast_period,
+            self.training_length,
+        )
+        self.assertEqual(len(result), 2)
+        self.assertCubeEqual(result[0], self.expected_period_forecast)
+        self.assertCubeEqual(result[1], self.expected_period_truth)
+
+    def test_duplicate_row_truths(self):
+        """Test that a truth cube is still produced if a single
+        duplicated truth for a given validity time is provided."""
+        truth_df_with_duplicates = self.truth_subset_df.append(
+            self.truth_subset_df.iloc[0], ignore_index=True
+        )
+        result = forecast_and_truth_dataframes_to_cubes(
+            self.forecast_df,
+            truth_df_with_duplicates,
+            self.cycletime,
+            self.forecast_period,
+            self.training_length,
+        )
+        self.assertEqual(len(result), 2)
+        self.assertCubeEqual(result[0], self.expected_period_forecast)
+        self.assertCubeEqual(result[1], self.expected_period_truth)
```

Comment:
What was the motivation for testing with both multiple and single duplicates? I'd rather see the single tests cover something slightly different, i.e. 3 identical entries to ensure the pandas drop functionality leaves you with just one.

There is also no checking here that the expected of two/three duplicates is returned, i.e. entries that share `["diagnostic", "forecast_period", "percentile", "time", "wmo_id"]` but have a different forecast value. This is not likely, but given that you've had to choose to keep the last of the entries, I think a test should check that what you get is the last of the entries.

Reply:
In practice, I would expect a situation more similar to `test_duplicate_cycle_forecasts` or `test_duplicate_cycle_truths` where a whole cycle/validity time is duplicated. I've included the `test_duplicate_row_forecasts` and `test_duplicate_row_truths` because these tests highlighted the need for the sorting that I added in this commit: https://github.com/metoppv/improver/pull/1614/commits/5c3b8c9b8b5b590ade1f5c17614af3b2fbe0a460, as this can end up with WMO IDs being in an unexpected order (i.e. "03003", "03004, "03002" for one blend time and "03002", "03003", "03004" for all other blend times). This subsequently leads to the cubes failing to merge together. I've modified the tests slightly to simplify them but I still want to keep multiple blend times and WMO_IDs within the input dataframe, so that it demonstrates the situation where the sorting is required. I've modified the "forecast" and "ob_value" within the rows with duplicates to demonstrate that the last entry is kept.